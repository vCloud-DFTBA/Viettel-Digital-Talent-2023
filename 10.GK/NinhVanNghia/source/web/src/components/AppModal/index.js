import { Modal, Button } from "react-bootstrap";


function AppModal({show, setShow, children, formik={}, title}) {
    return (
        <Modal data-testid="app-modal"
            show={show} onHide={() => setShow(false)}
            backdrop={true} keyboard={false}
            size={"lg"}
            centered
        >
            <Header setShow={setShow} title={title}/>
            <Modal.Body>
                {children}
            </Modal.Body>
            <Footer setShow={setShow} formik={formik}/>
        </Modal>
    )
}


export default AppModal


// ------------------------------------------------------------------


function Header({setShow, title}) {
    return (
        <Modal.Header style={{ height: "45px" }}>
            <Modal.Title>{title}</Modal.Title>
            <div
                className="btn btn-xs btn-icon btn-light cursor-pointer"
                onClick={() => setShow(false)}
            >
                X
            </div>
        </Modal.Header>
    )
}


function Footer({setShow, formik}) {
    return (
        <Modal.Footer>
            <div className='w-100 d-flex flex-row justify-content-around'>
                <Button className='font-weight-bold' variant="secondary" onClick={() => setShow(false)}>
                    Hủy bỏ
                </Button>
                <Button className='font-weight-bold' variant="primary" onClick={formik.handleSubmit} disabled={formik.isSubmitting}>
                    Thực hiện
                </Button>
            </div>
        </Modal.Footer>
    )
}