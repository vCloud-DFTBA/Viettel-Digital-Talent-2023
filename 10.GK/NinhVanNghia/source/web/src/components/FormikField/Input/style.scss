.FormikInput, .SelectInput {
    outline: none;
    p {
        height: 14px;
    }
}
.SelectInput{
    &:has(input:disabled) {
        background: #EFEFEF;
    }
}
