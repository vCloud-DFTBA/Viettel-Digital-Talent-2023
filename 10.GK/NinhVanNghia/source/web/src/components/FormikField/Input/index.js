import './style.scss'


export function Input({formik={}, name, ...rest}) {
    const {values, errors, touched, handleBlur, handleChange} = formik;
    return (
        <>
            <div className="bg-white border-bottom rounded px-2">
                <input
                    className="FormikInput Content_13 w-100 border-0 bg-transparent"
                    onChange={handleChange}
                    onBlur={handleBlur}
                    value={values?.[name]}
                    {...rest}
                />
            </div>
            <p className="text-danger Content_12">
                &nbsp;&nbsp;{touched?.[name] && errors?.[name] && errors?.[name]}
            </p>
        </>
    )
}


export function SelectInput({formik={}, name, options, ...rest}) {
    const {values, errors, touched, handleBlur, handleChange} = formik;
    const value = options?.findIndex(({value}) => value === values?.[name]) > -1
        ? values?.[name]
        : '' && (values[name] = '')
    return (
        <>
            <select className="SelectInput w-100 bg-white border rounded p-2"
                onChange={handleChange}
                onBlur={handleBlur}
                value={value}
                {...rest}
            >
                <option value='' label={`-- Select --`}/>
                {options?.map(({value, text}) => <option value={value} label={text} key={value}/>)}
            </select>
            <p className="text-danger Content_12">
                &nbsp;&nbsp;{touched?.[name] && errors?.[name] && errors?.[name]}
            </p>
        </>
    )
}

