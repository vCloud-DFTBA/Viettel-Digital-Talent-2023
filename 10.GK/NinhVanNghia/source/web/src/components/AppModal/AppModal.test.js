import renderer from 'react-test-renderer';
import userEvent from '@testing-library/user-event';
import { render, screen } from '@testing-library/react';

import AppModal from '.';
import * as Yup from 'yup';
import { useFormik } from 'formik';
import { Input } from "components/FormikField/Input";


test("It should render", async () => {
    expect(renderer.create(<AppModal/>).toJSON()).toMatchSnapshot()
})


test("It should submit form", async () => {
    const mockSubmit = jest.fn()

    // Wrap in function component to call useFormik hook
    function Wrapper() {
        const mockFormik = useFormik({
            initialValues: {
                username: '',
                fullname: '',
            },
            validationSchema: Yup.object({
                username: Yup.string().required('Bắt buộc').matches(/^[A-Za-z0-9]+$/, 'username chỉ chứa chữ, chữ hoa và số'),
                fullname: Yup.string().required('Bắt buộc').trim(),
            }),
            onSubmit: (values) => mockSubmit(values),
        })
        return (
            <AppModal show={true} formik={mockFormik}>
                <Input id="username" name="username" placeholder="username_test" formik={mockFormik}/>
                <Input id="fullname" name="fullname" placeholder="fullname_test" formik={mockFormik}/>
            </AppModal>
        )
    }

    render(<Wrapper/>)

    // Filling form
    await userEvent.type(screen.getByPlaceholderText('username_test'), username)
    await userEvent.type(screen.getByPlaceholderText('fullname_test'), fullname)

    // Submit form
    await userEvent.click(screen.getByRole('button', {name: "Thực hiện"}))

    expect(mockSubmit).toHaveBeenCalledWith({username, fullname})
})


// ---------------------------------------------------------


const username = 'anv'
const fullname = 'Nguyễn Văn A'
const birthDate = 2001
const university = 'BKHN'
const major = 'CNTT'
