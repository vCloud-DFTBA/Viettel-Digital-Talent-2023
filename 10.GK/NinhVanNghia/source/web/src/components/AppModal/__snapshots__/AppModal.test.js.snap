// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`It should render 1`] = `null`;
